### PR TITLE
Large refactor of `spin watch` to only run what is necessary when files change

### LINKED ISSUE #1403: `spin watch` loops infinitely if the `component.build` command modifies anything in `component.files`
By default `spin watch` watches for changes to `component.files` ([docs](https://developer.fermyon.com/common/cli-reference#watch)). If the `component.build.command` modifies these files then `spin watch` falls into an infinite loop. It is very typical for JS applications to have a build command that modifies the static files.

I see two potential solutions here:

1. Stop watching `component.files` when you run `spin watch` and only watch it when you run `spin watch --skip-build`. Passing the `--direct-mounts` flag can mitigate a lot of the downside here.
2. Differentiate between changes to files vs source code and dynamically change the underlying command that `spin watch` is running. When source code changes run `spin build --up`. When files change run `spin up`. (If we can get this to work this is my preferred approach.)

CC @itowlson I'd love to hear your thoughts here.

### PR DESCRIPTION
**Pull request by calebschoepp**
## Changes
- Large refactoring of code to clean things up and make it more clear
- `spin watch` now dynamically generates the command it will run. In practice what this means is that `spin build` is only run when actually necessary and we no longer get infinite loops.
- Closes #1403 .

## Notes
- The mental model you should have is that:
    - Watchexec produces `events` when things happen
    - Via debouncing these `events` are grouped into `actions`
    - These `events` from an `action` are reduced into a single `effect`
    - Based on the current `state` an `effect` produces an `outcome` and potentially changes the `state`
    - The `outcome` is passed back through the `action` to affect the world
- I'm planning to dump this into a README when I migrate this stuff to a crate

## Future work
- I feel like this command is really growing in scope -- as such I'm going to move it to its own crate in a future PR.
- I'm going to get live reconfiguration from the manifest working in a future PR.

### CODE REVIEW

File: src/watch_state.rs
Comment (lann) [inline src/watch_state.rs:73]: This might be a little more readable written as e.g.:

```suggestion
        let outcome = match (*state, effect) {
            (_, Effect::Exit) => {
                Outcome::if_running(Outcome::both(Outcome::Stop, Outcome::Exit), Outcome::Exit)
            },
            (_, Effect::ProcessFailed) => Outcome::if_running(Outcome::Stop, Outcome::DoNothing),
            (State::Building, Effect::ProcessCompleted) => {
                *state = State::Running;
                self.restart_outcome(true)
            },
            (State::Building, Effect::ProcessCompleted) => Outcome::DoNothing,
            ...
        };
```

File: src/watch_state.rs
Comment (lann) [inline src/watch_state.rs:98]: Took me a minute to understand what was going on here. Would this work?

```suggestion
        Outcome::sequence([
            Outcome::if_running(Outcome::Stop, Outcome::DoNothing),
            match should_clear {
                true => Outcome::Clear,
                false => Outcome::DoNothing,
            },
            Outcome::Start,
        ].into_iter())
```

File: src/commands/watch.rs
Comment (lann) [inline src/commands/watch.rs:111]: This is pretty subjective, but you might want to consider making the internal `State` private and exposing just what you need here as e.g. `state_machine.is_building() -> bool`

File: src/watch_filter.rs
Comment (lann) [inline src/watch_filter.rs:249]: I get the motivation for doing this, but given that these are only called inside boolean conditional expressions I'm pretty skeptical that this is a net positive. A single e.g. `WatchPattern::new(app_dir, glob, active)` method would cut quite a bit of code and probably be easier to read overall.

File: src/commands/watch.rs
Comment (lann) [inline src/commands/watch.rs:162]: I like the use of `Effect` ordering to choose the highest priority effect, but I wonder if you could replace `Effects` with a `Vec<Effect>` and do something like this here:

```suggestion
                // Pick the highest priority effect to apply
                let effect = effects.into_iter().min();
                action.outcome(state_machine.handle(effect));
```

The `Effects` abstraction is nice, but its only used in this one place so maybe not worth the extra code and indirection.

File: src/watch_filter.rs
Comment (lann) [inline src/watch_filter.rs:228]: This is what `#[derive(Debug)]` does internally (minus the one field):

```suggestion
        f.debug_struct("WatchPattern")
            .field("active", &self.active)
            .field("glob", &self.glob)
            .finish()
```

**Review by lann — APPROVED** [review]
Bunch of style/complexity suggestions, all optional.

File: src/commands/watch.rs
Comment (calebschoepp) [reply src/commands/watch.rs:239]: After lots of handholding from Clippy I landed on:

```rust
let sources = build.workdir.clone().map(|workdir| watch.iter()
                        .filter_map(|w| workdir.join(w).to_str().map(String::from))
                        .collect()).unwrap_or(watch);
```

File: src/commands/watch.rs
Comment (calebschoepp) [reply src/commands/watch.rs:111]: Leaving this one as is if only b/c I make heavy use of `get_state` in the tests.

Comment (lann) [reply src/commands/watch.rs:111]: Its not just about removing code. If you add another internal state like `BuildingAndDownloading` or something, this code (possibly) doesn't actually care about that new state; it just cares about some boolean property of that state.

Comment (lann) [reply src/commands/watch.rs:111]: (all of this with the caveat that it definitely isn't very important in the current codebase; this is just design navel-gazing)

File: src/commands/watch.rs
Comment (calebschoepp) [reply src/commands/watch.rs:230]: If a component doesn't have a build section then it will never get to this warning. AFAICT it is reasonable to warn with this on any component that has a build section.

File: src/commands/watch.rs
Comment (lann) [reply src/commands/watch.rs:239]: Yeah sorry; I should use an editor when writing some of these suggestions... :grimacing:

File: src/commands/watch.rs
Comment (lann) [reply src/commands/watch.rs:230]: I guess its a little annoying for my pet use-case of watching just the `source`. If I want to suppress this warning I have to copy `source` into `watch`.

File: src/commands/watch.rs
Comment (calebschoepp) [reply src/commands/watch.rs:162]: Think I'm going to keep this one. I like making the dependency on the ordering very explicit and built into `Effects`. It also means that we can make sure that this ordering isn't subtly broken by unit testing `Effects`.

File: src/watch_filter.rs
Comment (calebschoepp) [reply src/watch_filter.rs:228]: Thanks. Keep the hits coming 😄 . I'm learning a lot about Rust in these code reviews.

File: src/watch_state.rs
Comment (lann) [inline src/watch_state.rs:147]: ...in that case:

```suggestion
    pub fn reduce(&self) -> Effect {
        let effect = *self.0.min();
```

Comment (lann) [reply src/watch_state.rs:147]: Hmm...`min` isn't actually a method...what did I forget here?

File: src/watch_state.rs
Comment (calebschoepp) [reply src/watch_state.rs:98]: Cool. I never noticed `Outcome::sequence`

File: src/watch_state.rs
Comment (calebschoepp) [reply src/watch_state.rs:8]: I'm open to changing the name here but I don't think `WatchState` is better and I can't think of something. I'm trying to convey that it is a container tracking the current state and managing transitions.

Comment (lann) [reply src/watch_state.rs:8]: > I'm trying to convey that it is a container tracking the current state and managing transitions.

`WatchState` sounds right to me :shrug:

File: src/commands/watch.rs
Comment (calebschoepp) [reply src/commands/watch.rs:229]: Now I understand. I get why that could be annoying for that use case. I'll file an issue.

File: src/commands/watch.rs
Comment (lann) [inline src/commands/watch.rs:265]: I suspect you meant to do:

```suggestion
                Some(WatchPattern::new(path_str.to_owned(), watch_this_source, app_dir.as_path()))
```

File: src/commands/watch.rs
Comment (lann) [inline src/commands/watch.rs:260]:
```suggestion
                let path_str = path.to_str()?;
```

**Review by lann — APPROVED** [review]

File: src/watch_filter.rs
Comment (itowlson) [inline src/watch_filter.rs:209]: 1000000% banned from calling this `Config`.  We already have 800 things called `Config` and I can no longer keep track of them.  I'd suggest something like `WatchPatterns` or `WhatToWatch` maybe?

Comment (itowlson) [inline src/watch_filter.rs:209]: I am a bit surprised that the docs thing is not nagging about things being public without doc comments.  I'm not too fussed but it will certainly be needed when you crate it.

File: src/watch_state.rs
Comment (itowlson) [reply src/watch_state.rs:8]: Unresolving this to agree that `StateMachine` is insufficiently descriptive.  Even just `WatchStateMachine` would help.

File: src/watch_state.rs
Comment (itowlson) [inline src/watch_state.rs:17]: This would be more readable (to me) if you pulled the match out into a `let initial_state = match...`.

File: src/watch_state.rs
Comment (itowlson) [inline src/watch_state.rs:104]: Reading where this was consumed I wasn't sure what "process" it referred to, whether some internal procedure or an actual system process.  Maybe rename to `BuildCompleted` or `ChildProcessCompleted`?

File: src/watch_state.rs
Comment (itowlson) [inline src/watch_state.rs:102]: Consider `ChildProcessFailed` for a bit more clarify at the call site?  I realise it is more verbose though.

File: src/watch_state.rs
Comment (itowlson) [inline src/watch_state.rs:36]: I really like the clarity (and testability) of the state machine approach.  Very nice.

File: src/watch_state.rs
Comment (itowlson) [inline src/watch_state.rs:40]: I am struggling here a bit with `Outcome::if_running` and `State::Running`.  Does `if_running` refer to the `State`, or to something that is happening within the state machine itself?

File: src/watch_state.rs
Comment (itowlson) [inline src/watch_state.rs:69]: The boolean here is not clear at the call site (I was reading the state machine going "what are all these false values"). It may seem a bit laborious but what about an enum a la `Clear::Allow` / `Clear::Prevent`?

File: src/watch_filter.rs
Comment (itowlson) [inline src/watch_filter.rs:19]: Consider keeping a mention of "watch" in the name.

File: src/watch_filter.rs
Comment (itowlson) [inline src/watch_filter.rs:97]: This checks `wp.active` whereas the earlier `matches_*_pattern` functions didn't.  Is this intentional?  If so, consider a comment on why; if not, maybe refactor so this calls the `matches_*_pattern` functions so they are guaranteed to stay in sync.

File: src/watch_filter.rs
Comment (itowlson) [inline src/watch_filter.rs:106]: I like how you subtly changed the name of the lambda argument to avoid `cargo fmt` messing up the code structure.  _shakes puny fist at `cargo fmt`_

File: src/watch_filter.rs
Comment (itowlson) [inline src/watch_filter.rs:167]: Is this "wrong" or just "we don't care about it"?

File: src/watch_filter.rs
Comment (itowlson) [inline src/watch_filter.rs:217]: When reading both the consuming code and the generation code, I found it a bit hard to follow what `active` was doing.  I got the impression that `active = false` meant "don't watch this," and that it was a permanent state, i.e. not temporarily moved inactive then reactivated again.  But if that's the case, why have the pattern at all?  Why not just leave it out?  So I'm clearly missing something (perhaps relating to how `active` is sometimes checked and sometimes not).  It would be good to comment the `active` field and/or rename it to make the intended use clearer.

File: src/commands/watch.rs
Comment (itowlson) [inline src/commands/watch.rs:224]: I'd strongly encourage you to move this lambda out to a named function ~so I can understand it~ to more visibly express intent.

File: src/commands/watch.rs
Comment (itowlson) [inline src/commands/watch.rs:272]: Consider giving the files iterator a name so you can write something like `let artifact_patterns = concat(source_patterns, files_patterns)`.

File: src/commands/watch.rs
Comment (itowlson) [inline src/commands/watch.rs:273]: Surely we can't need to mutate the application manifest here?!

**Review by itowlson — CHANGES_REQUESTED** [review]
Looks good (despite the comments - mostly stylistic nits I think) - thanks!

File: src/watch_filter.rs
Comment (lann) [reply src/watch_filter.rs:209]: Err...this is a very common pattern in rust packages...

Comment (lann) [reply src/watch_filter.rs:209]: The module isn't public, so the struct isn't crate-public.

Comment (lann) [reply src/watch_filter.rs:209]: To be more explicit: I think this is an appropriate name and certainly shouldn't be banned. Naming a type `Config` - like `Result`, `Error`, `Iter`, `Entry` - is a well-established convention. They can be differentiated by consumers by renaming (`use std::io::Error as IoError`) or module path (`watch_filter::Config`).

As a blatant appeal to authority, `wasmtime` appears to have 6 `struct Config`s: https://github.com/search?q=repo%3Abytecodealliance%2Fwasmtime+%22struct+Config%22&type=code

File: src/watch_state.rs
Comment (calebschoepp) [reply src/watch_state.rs:8]: Two's a pattern. Going to go with `WatchState`

File: src/watch_state.rs
Comment (calebschoepp) [reply src/watch_state.rs:36]: Thanks. It was really fun to design and code.

File: src/watch_state.rs
Comment (calebschoepp) [reply src/watch_state.rs:69]: I was debating if this deserved an enum and I decided it didn't. On second thought I agree that it does.

File: src/watch_filter.rs
Comment (calebschoepp) [reply src/watch_filter.rs:19]: Since I'm planning to move this to it's own crate I'm going to keep the name as just `Filter` since it is implied that it is for watching -- given the crate.

File: src/commands/watch.rs
Comment (calebschoepp) [reply src/commands/watch.rs:273]: `append` expects a `&mut`. We're taking a mutable reference to the collected vector `collect::<Result<Vec<WatchPattern>>>()`.

File: src/commands/watch.rs
Comment (calebschoepp) [reply src/commands/watch.rs:224]: Done

File: src/watch_state.rs
Comment (calebschoepp) [reply src/watch_state.rs:69]: I actually opted for consts. https://stackoverflow.com/questions/52545734/is-there-a-way-to-use-a-boolean-as-an-underlying-type-for-an-enum.

I felt mapping between bool and an enum was a little clunky.

File: src/watch_state.rs
Comment (calebschoepp) [reply src/watch_state.rs:104]: I like it.

File: src/watch_state.rs
Comment (calebschoepp) [reply src/watch_state.rs:102]: I like this too.

File: src/watch_state.rs
Comment (calebschoepp) [reply src/watch_state.rs:40]: `Outcome::if_running` refers to [this](https://docs.rs/watchexec/latest/watchexec/action/enum.Outcome.html). Notably it says:

> Logic against the state of the command should be expressed using these variants, rather than inside the action handler, as it ensures the state of the command is always the latest available when the outcome is executed.

`State::Running` means we believe that `spin up` is currently running or at least we want it to be.

Basically wrapping all the outcomes in `Outcome::if_running` protects us from a scenario in which somehow our internal state gets out of sync with the actual state of watchexec. Likely not to matter but good protection regardless.

I added a comment to try and add some clarity.

File: src/watch_filter.rs
Comment (calebschoepp) [reply src/watch_filter.rs:217]: Good call out. `active` is in fact permanent. A better name is `should_watch`. The reason we need this field is because we want to have all possible patterns in the filter so that we can match them in the `matches_*_pattern` methods. However, we don't always want to pay attention to changes in these patterns hence `should_watch`.

Comment (calebschoepp) [reply src/watch_filter.rs:217]: Renamed and commented.

File: src/watch_filter.rs
Comment (calebschoepp) [reply src/watch_filter.rs:97]: This is on purpose. Cleared up in code by renaming `active` and adding more comments.

**Comment by calebschoepp** [comment]
@itowlson I've addressed your changes. Looking for a re-review.

File: src/commands/watch.rs
Comment (itowlson) [reply src/commands/watch.rs]: Doh, that makes sense - thanks!  But I appreciate you refactoring it away anyway so I don't come across it in six months and ask all over again!

**Review by itowlson — APPROVED** [review]
I know we have an open question about whether `should_watch` is needed, but the rest of it looks great.  Approving so you can merge during my weekend (if you want to) regardless of the outcome of your `should_watch` investigation - thanks, this is good stuff!

**Comment by calebschoepp** [comment]
> I know we have an open question about whether `should_watch` is needed, but the rest of it looks great. Approving so you can merge during my weekend (if you want to) regardless of the outcome of your `should_watch` investigation - thanks, this is good stuff!

I ended up pulling it out

**Pull request merged**
